gossipdns/dns.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/gorilla/mux"
+	"github.com/miekg/dns"
+
+	. "github.com/weaveworks/weave/common"
+	"github.com/weaveworks/weave/net/address"
+)
+
+const (
+	topDomain        = "."
+	reverseDNSdomain = "in-addr.arpa."
+	etcResolvConf    = "/etc/resolv.conf"
+	udpBuffSize      = uint16(4096)
+)
+
+type DNSServer struct {
+	ns     *Nameserver
+	domain string
+	ttl    uint32
+	port   int
+
+	servers   []*dns.Server
+	upstream  *dns.ClientConfig
+	tcpClient *dns.Client
+	udpClient *dns.Client
+}
+
+func NewDNSServer(ns *Nameserver, domain string, port int, ttl uint32,
+	clientTimeout time.Duration) (*DNSServer, error) {
+
+	s := &DNSServer{
+		ns:        ns,
+		domain:    domain,
+		ttl:       ttl,
+		port:      port,
+		tcpClient: &dns.Client{Net: "tcp", ReadTimeout: clientTimeout},
+		udpClient: &dns.Client{Net: "udp", ReadTimeout: clientTimeout, UDPSize: udpBuffSize},
+	}
+	var err error
+	if s.upstream, err = dns.ClientConfigFromFile(etcResolvConf); err != nil {
+		return nil, err
+	}
+
+	err = s.listen(port)
+	return s, err
+}
+
+func (d *DNSServer) String() string {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "WeaveDNS (%s)\n", d.ns.ourName)
+	fmt.Fprintf(&buf, "  listening on port %d, for domain %s\n", d.port, d.domain)
+	fmt.Fprintf(&buf, "  response ttl %d\n", d.ttl)
+	return buf.String()
+}
+
+func (d *DNSServer) listen(port int) error {
+	udpListener, err := net.ListenPacket("udp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return err
+	}
+	udpServer := &dns.Server{PacketConn: udpListener, Handler: d.createMux(d.udpClient)}
+
+	tcpListener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		udpServer.Shutdown()
+		return err
+	}
+	tcpServer := &dns.Server{Listener: tcpListener, Handler: d.createMux(d.tcpClient)}
+
+	d.servers = []*dns.Server{udpServer, tcpServer}
+	return nil
+}
+
+func (d *DNSServer) ActivateAndServe() {
+	for _, server := range d.servers {
+		go func(server *dns.Server) {
+			server.ActivateAndServe()
+		}(server)
+	}
+}
+
+func (d *DNSServer) Stop() error {
+	for _, server := range d.servers {
+		if err := server.Shutdown(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func badDNSRequest(r *dns.Msg, w dns.ResponseWriter) {
+	m := dns.Msg{}
+	m.SetReply(r)
+	m.RecursionAvailable = true
+	m.Rcode = dns.RcodeNameError
+
+	if err := w.WriteMsg(&m); err != nil {
+		Log.Infof("[gossipdns] error responding: %v", err)
+	}
+}
+
+func (d *DNSServer) createMux(client *dns.Client) *dns.ServeMux {
+	m := dns.NewServeMux()
+	m.HandleFunc(d.domain, d.handleLocal)
+	m.HandleFunc(reverseDNSdomain, d.handleReverse(client))
+	m.HandleFunc(topDomain, d.handleRecursive(client))
+	return m
+}
+
+func (d *DNSServer) handleLocal(w dns.ResponseWriter, req *dns.Msg) {
+	Log.Debugf("[gossipdns] request: %+v", *req)
+	if len(req.Question) != 1 || req.Question[0].Qtype != dns.TypeA {
+		badDNSRequest(req, w)
+		return
+	}
+
+	hostname := req.Question[0].Name
+	addrs := d.ns.Lookup(hostname)
+
+	response := dns.Msg{}
```

Comment:
I think it would keep the code tidier by getting rid of these low level things. The DNS server should deal with higher level stuff, not with DNS messages fields...